### PR TITLE
Remove pyplot strong dependency

### PR DESCRIPTION
**Pull request by PaulDebus**
As discussed in #37 we could also move the plotting into an extension and make PyPlot a weak dependency. This PR tries this instead of commenting out all mentions of PyPlot.

It should be working well enough, two minor things remain:
- changes the default for `nonmaxsuppts` to not plot the result to work in the default case without PyPlot available
- `ransacdemo.jl` is still not ported, maybe I can get to it at some point

@nlw0 what do you think about this?

Cheers Paul

### CODE REVIEW

**Comment by PaulDebus** [comment]
Julia LTS (1.4) fails because it does not know about extensions. I naively thought it would be kind of backward compatible, but it apparently is not... I will look into this

**Comment by peterkovesi** [comment]
I would suggest changing CI to only test Julia 1.6 and later as 1.6 was the last LTS

**Comment by PaulDebus** [comment]
The [packaging documentation](https://pkgdocs.julialang.org/v1/creating-packages/#Transition-from-normal-dependency-to-extension) shows a way to move code into an extension while keeping backwards compatibility:

1. Add the `PyPlot` dependency back into `[deps]`, which will be ignored in julia 1.9+
2. Add the following snippet to the main file
```julia
if !isdefined(Base, :get_extension)
  include("../ext/ImageProjectiveGeometryPyPlotExt.jl")
end
```

While this should work, we keep the strong dependency for older Julia<1.9. This keeps the CI complicated and requires projects that want slim dependencies to use at least Julia 1.9+. 

The alternative would be to only support new Julia versions, but given that 1.6 is the LTS, this is probably not a good idea...

**Comment by nlw0** [comment]
Sorry I couldn't come back to this yet. I'm not familiar yet with the `weakdep` mechanism, but seems promising. I'd say let's just focus on 1.9 moving forward, better to make this a great package depending on modern Julia than to worry about older versions. I'm myself still requiring 1.8 for something, but depending on 1.9 for future versions seems fine to me.

Now that's assuming this is a small fix. Ideally I'd say we really make this a very "mathematical" package, move anything with plotting somewhere else (maybe keep on the same repo, although I'd guess it's easier for julia packages to keep it simple). Now there's just no plotting anywhere, and we retain support for older versions.

File: src/ImageProjectiveGeometry.jl [hunk 1]
```diff
@@ -164,8 +164,8 @@ include("transforms3d.jl")
 include("ransac.jl")
 include("cornerfeatures.jl")
 include("utilities.jl")
-include("ransacdemo.jl")
+# include("ransacdemo.jl")
 include("geometry.jl")
+include("plotting.jl")
```
Comment (nlw0) [inline src/ImageProjectiveGeometry.jl:168]: Shouldn't this include go now inside the extension module?

File: src/plotting.jl [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+or more cameras showing their pose.  This consists of a rectangular cone,
+with its vertex at the camera centre, indicating the camera's field of view.
+The camera's coordinate X and Y axes are also plotted at the camera centre.
+
+See also: Camera
+"""
+function plotcamera(Ci, l; col=[0,0,1], plotCamPath=false, fig=nothing)
+    ext = get_plot_backend()
+    return ext.plotcamera(Ci, l, col, plotCamPath, fig)
+end
```
Comment (nlw0) [inline src/plotting.jl]: Sorry for nitpicking, please configure your editor to add newlines at the end of files.

File: src/ImageProjectiveGeometry.jl [hunk 3]
```diff
@@ -164,8 +164,8 @@ include("transforms3d.jl")
 include("ransac.jl")
 include("cornerfeatures.jl")
 include("utilities.jl")
-include("ransacdemo.jl")
+# include("ransacdemo.jl")
```
Comment (nlw0) [inline src/ImageProjectiveGeometry.jl]: Maybe this could be simply included in the extension module as well (see next comment)

**Comment by nlw0** [comment]
I'd like to try the package out first, but I'm thinking maybe we go ahead with this. Let me get this straight, this keeps compatibility with 1.6, even if CI breaks or whatever, and any 1.6 users might figure out what to do, correct? We merely gain a "new feature" on 1.9, and are able to ignore PyPlot in CI.

I'm thinking this would look like and API change, and perhaps we should now bump a major version? Maybe not now, I guess we can finish the PR first and see if we want to make any other big changes.

File: src/plotting.jl (hunk 2 above)
Comment (PaulDebus) [reply src/plotting.jl]: Good point, not nitpicky at all.

File: src/ImageProjectiveGeometry.jl (hunk 1 above)
Comment (PaulDebus) [reply src/ImageProjectiveGeometry.jl:168]: Could be. I was thinking about allowing different plotting backends, as PyPlot can be quite tricky to set up, and I have used Makie recently, which worked quite well. 
I guess my idea was to have some logic that defines the interface of the plotting (which functions need to be implemented) and decides which plotting backend to use - and also is more explicit if somebody tries to use one of the functions.

Looking at the docs again, I think we could also move it into the extension module. 

I guess you do not care too much about plotting, so I could look into it and make sure, that all plotting backends have the correct interface.

File: src/ImageProjectiveGeometry.jl (hunk 3 above)
Comment (PaulDebus) [reply src/ImageProjectiveGeometry.jl]: yeah, maybe. It should go somewhere. I guess, we should add the move to the readme, as the demo would be only available if you take additional steps.

**Comment by PaulDebus** [comment]
So this would make a shiny feature in 1.9+, where we have the state we want: plotting works if you manually add PyPlot. If you do not add PyPlot, then no plotting is happening at all and CI for example would be easier (if we do not want to test the plotting code)

For everybody <1.9, a small change (see https://github.com/peterkovesi/ImageProjectiveGeometry.jl/pull/38#issuecomment-1651418989) is required, that adds PyPlot as a normal dependency (again). Then, nothing is different from now. 

Moving the plotting out of the default functionality would warrant a version change I guess. 

So things I want to change before merging:
- Make adjustments for Julia <1.9 to work again (see https://github.com/peterkovesi/ImageProjectiveGeometry.jl/pull/38#issuecomment-1651418989) 
- Move code from `plotting.jl` into the extension
- move the ransac demo into the extension
- update Readme to show the changes and give an explanation on how to get plotting back

This will require CI to still set up Pyplot for the LTS Julia (1.6), which is still broken (see the other discussion). 

I hope I did not forget anything... Would you be ok with all that?

**Comment by PaulDebus** [comment]
Hey everyone,

I put some more time into making this work and implemented the changes described above. This does not fundamentally change things, as tests are still broken for older Julia versions. But for Julia >=1.9, we do not need PyPlot anymore. 

As a next step, we could use [Requires.jl](https://github.com/JuliaPackaging/Requires.jl) to provide the same functionality, as is also mentioned in the [Julia packaging docs](https://pkgdocs.julialang.org/v1/creating-packages/#Backwards-compatibility). I initially did not like this solution, as it adds more dependencies, but the longer I think about it, the more I like it. The alternative would still be dropping support for Julia <1.9, but I do not want to decide this alone.

What do you say?

Best, Paul

**Comment by peterkovesi** [comment]
I personally see little problem with requiring Julia >= 1.9 especially if it makes package development simpler.
I have no idea how many users work with older versions of Julia but I have never encountered any problem when upgrading  a Julia version, and upgrading is a simple process.

**Comment by PaulDebus** [comment]
Ok, I also see no problem in requiring a newer Julia version. If this is fine for everybody, then the latest changes should be fine:
- require Julia >=1.9
- change CI tests to use Julia 1.9 and latest
- drop Requires.jl
- Bump package version to `0.4.0`
- Keep the not so nice `get_plot_backend` function to select the correct backend and route the function calls to the correct package extension. This avoids some method overwrite warnings while loading the extension. 

The tests seem to be fine except for the coverage. The plotting functions are not tested, so coverage is not increased and the coverage of the diff is small. 

I would appreciate it if somebody could take a look at the resulting changes, to check if I missed something.

Thanks and all the best
Paul

**Comment by PaulDebus** [comment]
Any comments on that? I would like to merge that to get it out of my head (and also simplify the testing for another package I am working on).

If there are no objections, I would merge this soon.

Best, Paul

**Review by peterkovesi — APPROVED** [review]
Sorry to have held you up.  I think it looks good to go. (Sorry about all my trailing spaces on so many lines of code!)

**Comment by PaulDebus** [comment]
No worries, I was also busy and did not keep track. 
Thanks for merging, should this trigger a new GitHub release?

**Comment by peterkovesi** [comment]
Yes I guess this should be a new release as a fair bit has changed.  Perhaps also add a brief note to the README explaining the plotting changes and the need for Julia 1.9 or greater.

**Comment by PaulDebus** [comment]
I had already added some information to the README in the Installation section. Do you think this is sufficient?

I can of course extend it a bit.

**Comment by peterkovesi** [comment]
What you have added to the README is fine.  Sorry, I should have looked first!

**Pull request merged**